packages/grid/_modules_/grid/hooks/features/columns/columnsUtils.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  getGridColDef,
+  getGridDefaultColumnTypes,
+  GRID_STRING_COL_DEF,
+  gridCheckboxSelectionColDef,
+  GridColDef,
+  GridColumnLookup,
+  GridColumns,
+  GridColumnTypesRecord,
+  GridColumnsState,
+  GridLocaleText,
+  GridStateColDef,
+  GridTranslationKeys,
+  Logger,
+} from '../../../models';
+import { mergeGridColTypes } from '../../../utils';
+
+export type RawGridColumnsState = Omit<GridColumnsState, 'lookup'> & {
+  lookup: { [field: string]: GridColDef | GridStateColDef };
+};
+
+export function getStateColumns(
+  columns: (GridColDef | GridStateColDef)[],
+  viewportWidth: number,
+): GridStateColDef[] {
+  let totalFlexUnits = 0;
+  let widthToAllocateInFlex = viewportWidth;
+
+  const stateColumns: GridStateColDef[] = [];
+
+  // Compute the width of non-flex columns and how much width must be allocated between the flex columns
+  for (let i = 0; i < columns.length; i += 1) {
+    const column = { ...columns[i] } as GridStateColDef;
+
+    if (column.hide) {
+      column.computedWidth = 0;
+    } else {
+      const minWidth = column.minWidth ?? GRID_STRING_COL_DEF.minWidth!;
+
+      if (column.flex && column.flex > 0) {
+        totalFlexUnits += column.flex;
+        column.computedWidth = minWidth;
+      } else {
+        const computedWidth = Math.max(column.width ?? GRID_STRING_COL_DEF.width!, minWidth);
+        column.computedWidth = computedWidth;
+        widthToAllocateInFlex -= computedWidth;
+      }
+    }
+
+    stateColumns.push(column);
+  }
+
+  // Compute the width of flex columns
+  if (totalFlexUnits && widthToAllocateInFlex > 0) {
+    const widthPerFlexUnit = totalFlexUnits > 0 ? widthToAllocateInFlex / totalFlexUnits : 0;
+
+    for (let i = 0; i < stateColumns.length; i += 1) {
+      const column = stateColumns[i];
+
+      if (!column.hide && column.flex && column.flex > 0) {
+        const flexColumnWidth = Math.floor(widthPerFlexUnit * column.flex);
```

Comment:
as for #2110, it creates issues to use the `Math.floor` here

Reply:
https://github.com/mui-org/material-ui-x/pull/2110/files#diff-ab3374e32d059e6ec4d957bf3002146d73129b88296f556ae19061e1c15512c6R63